compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -741,25 +729,374 @@ fn encode_ty<'tcx>(
+pub fn fn_trait_args<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    predicates: &List<ty::PolyExistentialPredicate<'tcx>>,
+) -> Option<&'tcx List<Ty<'tcx>>> {
+    if is_fn_trait(tcx, predicates) {
+        if let Some(args) = principal_args(predicates)
+            && !args.is_empty()
+            && args[0].expect_ty().is_tuple()
+        {
+            Some(args[0].expect_ty().tuple_fields())
+        } else {
+            bug!("fn_trait_args: unexpected non-tuple arg `{:?}`", principal_args(predicates));
+        }
+    } else {
+        None
+    }
+}
+
+/// Returns the output if principal is an Fn trait.
+pub fn fn_trait_output<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    predicates: &List<ty::PolyExistentialPredicate<'tcx>>,
+) -> Option<Ty<'tcx>> {
+    if is_fn_trait(tcx, predicates) {
+        if let Some(projection) = predicates.projection_bounds().next()
+            && let Some(ty) = projection.skip_binder().term.ty()
+        {
+            Some(ty)
+        } else {
+            bug!(
+                "fn_trait_output: unexpected non-term projection `{:?}`",
+                predicates.projection_bounds().next()
+            );
+        }
+    } else {
+        None
+    }
+}
+
+/// Returns the signature if principal is an Fn trait.
+pub fn fn_trait_sig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    predicates: &List<ty::PolyExistentialPredicate<'tcx>>,
+) -> Option<FnSig<'tcx>> {
+    if is_fn_trait(tcx, predicates) {
+        Some(tcx.mk_fn_sig(
+            fn_trait_args(tcx, predicates).unwrap(),
+            fn_trait_output(tcx, predicates).unwrap(),
+            false,
+            hir::Unsafety::Normal,
+            Abi::Rust,
+        ))
+    } else {
+        None
+    }
+}
+
+/// Returns true if the FnAbi is an unpacked "rust-call" FnAbi.
+fn is_closure_fnabi<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> bool {
+    if fn_abi.conv == Conv::Rust && !fn_abi.c_variadic && !fn_abi.args.is_empty() {
+        match fn_abi.args[0].layout.ty.kind() {
+            ty::RawPtr(ty::TypeAndMut { ty, .. }) | ty::Ref(_, ty, _) => {
+                // FIXME(rcvalle): Currently both fn() and fn(f: fn()) will have the same type ids.
+                //   Disambiguate between the possibly by passign down the "rust-call" ABI.
+                if let ty::Closure(_, args) = ty.kind() {
+                    let fn_sig =
+                        tcx.signature_unclosure(args.as_closure().sig(), hir::Unsafety::Normal);
+                    return fn_sig
+                        .skip_binder()
+                        .inputs()
+                        .iter()
+                        .eq_by(&fn_abi.args[1..], |ty, arg| *ty == arg.layout.ty)
+                        && fn_sig.output().skip_binder() == fn_abi.ret.layout.ty;
+                }
+            }
+            _ => {}
+        }
+    }
+    false
+}
+
+/// Returns true if the FnAbi is a dynamic Fn trait FnAbi.
+fn is_dynamic_fn_fnabi<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> bool {
```

Comment:
Can you explain what this does? As far as I can tell, this will also match *free* functions like:

```
fn foo(x: &&&&&dyn Fn(i32), x: i32) {}
```

Which seems not right.

Reply:
It identifies an unwrapped dynamic `Fn` call, but I shouldn't have used `peel_refs`. Thanks for pointing it out! I'll fix it.

Reply:
(FWIW, I think it may be possible to include information in the FnAbi to identify when it's a "rust-call" or an unwrapped dynamic Fn call.)

Reply:
Fixed the peel_refs, but I still need to disambiguate the dynamic Fn trait calls.

Reply:
It seems that after the `transform_fnabi`, dynamic Fn trait declarations/definitions are disambiguated as well. They match `ty::InstanceDef::Item(..)` and are not handled there, skipping the else block, and returning an unchanged FnAbi at the end. I added test cases for both closures and dynamic Fn trait disambiguations.